core/state/snapshot/generate.go

```diff
@@ -527,6 +542,15 @@ func (dl *diskLayer) generateRange(root common.Hash, prefix []byte, kind string,
 	logger.Debug("Regenerated state range", "root", root, "last", hexutil.Encode(last),
 		"count", count, "created", created, "updated", updated, "untouched", untouched, "deleted", deleted)
 
+	// Clean up the left dangling storages at the end in case the entire
+	// snapshot is regenerated. Note if the last is nil, it means there
+	// is no snapshot data at disk at all, the entire snapshot is expected
+	// to be re-generated already.
+	if last == nil && drange != nil {
+		if err := drange.cleanup(nil); err != nil {
+			return false, nil, err
+		}
```

Comment:
I see that you do a `cleanup`, but I don't see where you actually ever invoke `detect` ?

Reply:
detect is invoked when the range is constructed.